File: trust-sdk-client/src/main/java/trust/Trust.java
package trust;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;

public abstract class Trust {

    public static final String ACTION_SIGN_TRANSACTION = "sign-transaction";
    public static final String ACTION_SIGN_MESSAGE = "sign-message";
    public static final String ACTION_SIGN_PERSONAL_MESSAGE = "sign-personal-message";
    public static final String ACTION_SIGN_TYPED_MESSAGE = "sign-typed-message";

    public static final int RESULT_ERROR = 1;

    public static final int REQUEST_CODE_SIGN = 8001;

    private static String packageName = "com.wallet.crypto.trustapp";

    private Trust() {}

    public static void setWalletAppPackageName(String packageName) {
        Trust.packageName = packageName;
    }

    public static SignTransactionRequest.Builder signTransaction() {
        return SignTransactionRequest.builder();
    }

    public static SignMessageRequest.Builder signMessage() {
        return SignMessageRequest.builder();
    }

    public static SignPersonalMessageRequest.Builder signPersonalMessage() {
        return SignPersonalMessageRequest.builder();
    }

    public static SignTypedMessageRequest.Builder signTypedMessage() {
        return SignTypedMessageRequest.builder();
    }

    public static <T extends Request> Call<T> execute(final Activity activity, T request) {
        final Intent intent = new Intent();
        intent.setData(request.key());
        if (canStartActivity(activity, intent)) {
            activity.startActivityForResult(intent, REQUEST_CODE_SIGN);
            return new Call<>(request);
        } else {
            try {
                activity.startActivity(new Intent(
                        Intent.ACTION_VIEW,
                        Uri.parse("market://details?id=" + packageName)));
            } catch (android.content.ActivityNotFoundException anfe) {
                activity.startActivity(new Intent(
                        Intent.ACTION_VIEW,
                        Uri.parse("https://play.google.com/store/apps/details?id=" + packageName)));
            }
            return null;
        }
    }

    static boolean canStartActivity(Context context, Intent intent) {
        PackageManager pm = context.getPackageManager();
        return pm.queryIntentActivities(intent, 0).size() > 0;
    }

    interface ExtraKey {
        String SIGN = "sign";
        String ERROR = "error";
        String RECIPIENT = "to";
        String VALUE = "amount";
        String GAS_PRICE = "gasPrice";
        String GAS_LIMIT = "gasLimit";
        String CONTRACT = "contract";
        String PAYLOAD = "data";
        String NONCE = "nonce";
        String LEAF_POSITION = "leaf_position";
        String MESSAGE = "message";
        String URL = "url";
        String CALLBACK_URI = "callback";
    }

    public interface ErrorCode {
        int UNKNOWN = -1;
        int NONE = 0;
        int CANCELED = 1;
        int INVALID_REQUEST = 2;
        int WATCH_ONLY = 3;
    }
}

File: trust-sdk-client/src/main/java/trust/Trust.kt
package trust

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri

object Trust {
    const val RESULT_ERROR = 1
    var packageName = "com.wallet.crypto.trustapp"

    fun setWalletAppPackageName(packageName: String) {
        Trust.packageName = packageName
    }

    fun <R, T : Request<R>> execute(activity: Activity, request: T): Call<R, T>? {
        val intent = Intent()
        intent.data = request.data()
        return if (canStartActivity(activity, intent)) {
            activity.startActivityForResult(intent, request.getRequestCode())
            Call(request)
        } else {
            openMarket(activity)
            null
        }
    }

    fun execute(activity: Activity, operation: Operation) {
        val intent = Intent(Intent.ACTION_VIEW, operation.buildUri())
        if (canStartActivity(activity, intent)) {
            activity.startActivity(intent)
        } else {
            openMarket(activity)
        }
    }

    fun handleOperationResult(intent: Intent?): OperationResult? = when (intent?.data?.getQueryParameter(ExtraKey.ACTION.key)) {
        Action.TRANSFER.key,
        Action.TRADE.key,
        Action.DELEGATE.key,
        Action.GET_ACCOUNTS.key -> buildOperationResult(intent)
        else -> null
    }

    private fun buildOperationResult(intent: Intent?): OperationResult? {
        val signature = intent?.data?.getQueryParameter(ExtraKey.TRANSACTION_SIGN.key)
        val hash = intent?.data?.getQueryParameter(ExtraKey.TRANSACTION_HASH.key)
        val data = intent?.data?.getQueryParameter(ExtraKey.DATA.key)
        val error = intent?.data?.getQueryParameter(ExtraKey.CANCEL.key)
        val id = intent?.data?.getQueryParameter(ExtraKey.ID.key)
        val operationError = if (error.isNullOrEmpty()) null else OperationError.safeValueOf(error)

        return OperationResult(
            data = data ?: hash ?: signature,
            error = operationError,
            id = id?.toIntOrNull()
        )
    }

    private fun openMarket(activity: Activity) {
        try {
            activity.startActivity(Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("market://details?id=${packageName}")))
        } catch (ignored: ActivityNotFoundException) {
            activity.startActivity(Intent(
                    Intent.ACTION_VIEW,
                    Uri.parse("https://play.google.com/store/apps/details?id=${packageName}")))
        }
    }

    private fun canStartActivity(context: Context, intent: Intent): Boolean {
        val pm = context.packageManager
        return pm.queryIntentActivities(intent, 0).size > 0
    }

    internal enum class ExtraKey(val key: String) {
           ERROR("error"),
           COINS("coins"),
           ADDRESSES("addresses"),
           TRANSACTION_HASH("transaction_hash"),
           TRANSACTION_SIGN("transaction_sign"),
           DATA("data"),
           CANCEL("cancel"),
           ACTION("action"),
           ID("id"),
    }

    internal enum class Action(val key: String) {
        TRANSFER("transfer"),
        TRADE("trade"),
        DELEGATE("delegate"),
        GET_ACCOUNTS("get_accounts"),
    }

    internal enum class Host(val key: String) {
        SDK_GET_ACCOUNTS("sdk_get_accounts"),
        ACCOUNTS_CALLBACK("accounts_callback"),
        TX_CALLBACK("tx_callback"),
        SDK_TRANSACTION("sdk_transaction"),
    }
}